text/0102-offchain-parachain-runtime-upgrades.md

```diff
@@ -0,0 +1,347 @@
+
+## Summary
+
+Change the upgrade process of a parachain runtime upgrade to become an off-chain
+process with regards to the relay chain. Upgrades are still contained in
+parachain blocks, but will no longer need to end up in relay chain blocks nor in
+relay chain state.
+
+## Motivation
+
+Having parachain runtime upgrades go through the relay chain has always been
+seen as a scalability concern. Due to optimizations in statement
+distribution and asynchronous backing it became less crucial and got
+de-prioritized, the original issue can be found
+[here](https://github.com/paritytech/polkadot-sdk/issues/971).
+
+With the introduction of Agile Coretime and in general our efforts to reduce
+barrier to entry more for Polkadot more, the issue becomes more relevant again:
+We would like to reduce the required storage deposit for PVF registration, with
+the aim to not only make it cheaper to run a parachain (bulk + on-demand
+coretime), but also reduce the amount of capital required for the deposit. With
+this we would hope for far more parachains to get registered, thousands
+potentially even ten thousands. With so many PVFs registered, updates are
+expected to become more frequent and even attacks on service quality for other
+parachains would become a higher risk.
+
+## Stakeholders
+
+- Parachain Teams
+- Relay Chain Node implementation teams
+- Relay Chain runtime developers
+
+## Explanation
+
+The issues with on-chain runtime upgrades are:
+
+1. Needlessly costly.
+2. A single runtime upgrade more or less occupies an entire relay chain block, thus it
+   might affect also other parachains, especially if their candidates are also
+   not negligible due to messages for example or they want to uprade their
+   runtime at the same time.
+3. The signalling of the parachain to notify the relay chain of an upcoming
+   runtime upgrade already contains the upgrade. Therefore the only way to rate
+   limit upgrades is to drop an already distributed update in the size of
+   megabytes: With the result that the parachain missed a block and more
+   importantly it will try again with the very next block, until it finally
+   succeeds. If we imagine to reduce capacity of runtime upgrades to let's say 1
+   every 100 relay chain blocks, this results in lot's of wasted effort and lost
+   blocks.
+
+We discussed introducing a separate signalling before submitting the actual
+runtime, but I think we should just go one step further and make upgrades fully
+off-chain. Which also helps bringing down deposit costs in a secure way, as we
+are also actually reducing costs for the network.
+
+### Introduce a new UMP message type `RequestCodeUpgrade`
+
+As part of elastic scaling we are already planning to increase flexibility of [UMP
+messages](https://github.com/polkadot-fellows/RFCs/issues/92#issuecomment-2144538974), we can now use this to our advantage and introduce another UMP message:
```

Comment:
"We just need this hack for one thing and will not use it for anything else" ;)

Reply:
It does feels indeed it is creeping out into other features/change like this one but it offers a lot of advantages in the short term. I would not call it a hack, but more of a generalisation of the UMP queue. The alternative is PVF versioning which I believe is the long term solution that we'll likely to develop in 2025.

Reply:
I mean for `CoreIndex` it is clearly a hack. `RequestCodeUpgrade` is an actual message that is sort of fine to be passed here. However, this brings up the topic on, should we add it to XCM? Should we make UMP messages generic, where one variant is `XCM` and the others are more UMP related?

Reply:
We don't want to add it to XCM, instead we will have a UMP queue separator between regular XCM messages and the possible additional ones for `CoreIndex` and the `RequestCodeUpgrade`. I will soon post the RFC which explains the UMP changes in detail.

Reply:
https://github.com/polkadot-fellows/RFCs/pull/103

Reply:
> We don't want to add it to XCM, instead we will have a UMP queue separator between regular XCM messages and the possible additional ones for `CoreIndex` and the `RequestCodeUpgrade`

I know what the plan was/is. However, this doesn't really invalidate what I said above.

Reply:
Ah, I see. I would prefer to make the UMP messages more generic in this case, having two variants, one wrapping XCM and the other UMPSignal as defined here. Sounds much better than using a separator.